src/jsMain/kotlin/io/beatmaps/shared/form/multipleChoice.kt

```diff
@@ -0,0 +1,52 @@
+package io.beatmaps.shared.form
+
+import external.reactFor
+import kotlinx.html.InputType
+import kotlinx.html.id
+import kotlinx.html.js.onChangeFunction
+import org.w3c.dom.HTMLInputElement
+import react.PropsWithRef
+import react.RBuilder
+import react.RComponent
+import react.State
+import react.dom.div
+import react.dom.input
+import react.dom.label
+
+external interface MultipleChoiceProps<T> : PropsWithRef<HTMLInputElement> {
+    var name: String
+    var choices: Map<String, T>
+    var selectedValue: T
+    var block: ((T) -> Unit)?
+    var className: String?
+}
+
+class MultipleChoice<T> : RComponent<MultipleChoiceProps<T>, State>() {
```

Comment:
I guess I have the same question. Are we really getting any benefit from this approach, using functional components and making it less type safe in the process, over using a generic class component?

Reply:
As far as I'm aware all the current work in react is on functional components. All the examples use them, they're easier to read.

Where are you losing type safety? Functional components have the same typing as class components. You can wrap up any generics in their properties as I was trying to demonstrate.

Reply:
> Where are you losing type safety?

I don't know, that's just what you said.

I will try out this approach.

Reply:
What I was trying to demonstrate is your approach never had type safety. The block from my screenshot was always fed an `Any?` so in a sense you may as well just type it like that (which I don't like). But finding a way to bury the logic in classes where it can be maintained is not easy.